### PR TITLE
JSON LD branch

### PR DESCRIPTION
**Pull request by benosteen**
### Types of changes

-   [x] Bugfix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Description of Change

This branch adds in the JSON-LD export and import that has been worked on in parallel to the master branch. It addresses a number of issues, and while there are some outstanding issues, it is at a stage where incorporating the foundation of the work into the core branch is sensible.

### Issues Solved

https://github.com/archesproject/arches/issues?utf8=%E2%9C%93&q=is%3Aclosed+label%3A%22Roadmap%3A+Graph+Designer+Update+-+Task+1+%28API%29%22+

### Checklist

-   [x] Unit tests pass locally with my changes
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [ ] I have added necessary documentation (if appropriate)

### Further comments

Datatypes now control how they are represented in JSON-LD, by implementing the `to_rdf` and `from_rdf` methods. 

`to_rdf` takes objects representing this datatype in the Arches system and outputs an RDF graph (using `rdflib`'s `Graph` object) that represents this datatype's data.

`from_rdf` takes in the data structure that corresponds to a Datatype from a JSON-LD expanded document, and returns the data that this Datatype is required to store in the Arches instance.

### CODE REVIEW

File: arches/app/datatypes/datatypes.py [hunk 1]
```diff
@@ -1483,21 +1562,16 @@ def _add_resource(d, p, r, r_type):
...
-        # NOTE: This returns a dict including the host and the resource instance UUID as
-        # {"host":..., "res_inst":...}
-        # It may be important in future for resource instance datatypes to deal with externally held ones
-        # rather than require a local copy.
+        res_inst_uri = json_ld_node['@id']
+        # `id` should be in the form schema:{...}/{UUID}
+        # eg `urn:uuid:{UUID}`
+        #    `http://arches_instance.getty.edu/resources/{UUID}`
 
         import re
```
Comment (apeters) [inline arches/app/datatypes/datatypes.py]: can we remove this an add it at the top of the file

File: arches/app/views/api.py [hunk 2]
```diff
@@ -380,26 +389,34 @@ def post(self, request, resourceid=None):
...
                     for value in reader.errors.itervalues():
                         response.append(value.message)
-                    return JSONResponse(data, indent=indent, status=400, reason=response)
+                    return JSONResponse({"error": response}, indent=indent, status=400)
                 else:
                     response = []
                     for resource in reader.resources:
                         with transaction.atomic():
                             resource.save(request=request)
+                        print resource.resourceinstanceid
```
Comment (apeters) [inline arches/app/views/api.py]: we can remove the print statement

File: arches/app/views/api.py [hunk 3]
```diff
@@ -380,26 +389,34 @@ def post(self, request, resourceid=None):
...
                         response.append(JSONDeserializer().deserialize(
                             self.get(request, resource.resourceinstanceid).content))
                     return JSONResponse(response, indent=indent, status=201)
             else:
                 return JSONResponse(status=403)
         except Exception as e:
-            return JSONResponse(status=500, reason=e)
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            formatted = traceback.format_exception(exc_type, exc_value, exc_traceback)
+            if len(formatted):
```
Comment (apeters) [inline arches/app/views/api.py]: can we remove/comment this whole block?

File: arches/app/views/api.py (hunk 2 above)
Comment (benosteen) [reply arches/app/views/api.py]: Probably? It's from the master branch which I had to pull in to deal with the django migration issue that had got out of line. Shall I do this here?

Comment (benosteen) [reply arches/app/views/api.py]: Or maybe it was something Rob S put in... either way I can remove it.

Comment (benosteen) [reply arches/app/views/api.py]: Removed

File: arches/app/views/api.py (hunk 3 above)
Comment (benosteen) [reply arches/app/views/api.py]: I've wrapped the code as I suggested. Is this a suitable change?

File: arches/app/views/api.py (hunk 2 above)
Comment (azaroth42) [reply arches/app/views/api.py]: Yep, there's lots of debugging still in there. Happy to remove most of it.

File: arches/app/views/api.py (hunk 3 above)
Comment (azaroth42) [reply arches/app/views/api.py]: This one OTOH is more important. The response is meaningless without what actually happened.

Comment (azaroth42) [reply arches/app/views/api.py]: mode=DEV is a good solution, :+1: to @benosteen

**Comment by apeters** [comment]
@benosteen this PR looks good to me.  It's ready to merge into master.

**Pull request merged**